### PR TITLE
Enclose exported environment variable values with quotes, fixes #3499

### LINKED ISSUE #3499: variables in provider environment_variables don't get properly evaluated if the value has spaces in it
**Describe the bug**

String variables with space are not correctly transferred from yaml to bash.

**To Reproduce**

* Create a custom provider configuration
* Add a variable which contains a space in the value
* Use the variable in a command

```yaml
environment_variables:
  test: "variable with spaces"

auth_command:
  command: |
    set -eu -o pipefail

db_pull_command:
  command: |
    # set -x # You can enable bash debugging output by uncommenting
    set -eu -o pipefail
    echo ${test}

files_pull_command:
  command: |
    # set -x   # You can enable bash debugging output by uncommenting
    set -eu -o pipefail
    echo ${test}
  service: web
```

Run with `ddev pull test -y --skip-import`

Result:

```bash
Authenticating... 
Obtaining database... 
variable
Skipping database import. 
Obtaining files... 
variable
Skipping files import. 
Pull succeeded. 
```

**Expected behavior**

The complete string should be available in the bash commands

```bash
Authenticating... 
Obtaining database... 
variable with spaces
Skipping database import. 
Obtaining files... 
variable with spaces
Skipping files import. 
Pull succeeded. 
```

**Version and configuration information (please complete the following information):**

- Host macOS Monterey 12.1 Apple M1
- Docker version 20.10.11, build dea9396
- ddev version v1.18.2

**Additional context**
https://discord.com/channels/664580571770388500/912420298488447026/928257058099302421

### PR DESCRIPTION
**Pull request by DigitalFrontiersMedia**
## The Issue
Environment variables are exported inline with no delimiter between one value and the next variable.

## How This PR Solves The Issue
Encloses variable values with quotes to note beginning and end of values.

## Manual Testing Instructions

- Create a custom provider configuration

```
environment_variables:
  test: "variable with spaces"

auth_command:
  command: |
    set -eu -o pipefail

db_pull_command:
  command: |
    # set -x # You can enable bash debugging output by uncommenting
    set -eu -o pipefail
    echo ${test}

files_pull_command:
  command: |
    # set -x   # You can enable bash debugging output by uncommenting
    set -eu -o pipefail
    echo ${test}
  service: web
```

- Add a variable (as above) which contains a space in the value
- Echo the variable in a command (as above)
- Run with `ddev pull test -y --skip-import`

### Expected behavior
The complete string should be available in the bash commands

```
Authenticating... 
Obtaining database... 
variable with spaces
Skipping database import. 
Obtaining files... 
variable with spaces
Skipping files import. 
Pull succeeded. 
```

## Automated Testing Overview
None.

## Related Issue Link(s)
#3499

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/4594"></a>

<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4594"></a>

### CODE REVIEW

**Comment by rfay** [comment]
Looks like TestGitPull and maybe a thing or two other are broken; you should be able to run these locally or on gitpod pretty easily. https://ddev.readthedocs.io/en/latest/developers/building-contributing/#testing - if you don't have your vscode working with tests yet, then make sure to do that, it's way easy.

**Comment by DigitalFrontiersMedia** [comment]
I ran `make test` locally and there are still failures, but I don't understand how they have anything to do with what I'm doing.  What is the expectation here?

```
=== RUN   TestGetAppsEmpty
    ddevapp_test.go:2878: 
                Error Trace:    /Users/stephen/Documents/git/dfm-ddev/pkg/ddevapp/ddevapp_test.go:2878
                Error:          Not equal: 
                                expected: 0
                                actual  : 1
                Test:           TestGetAppsEmpty
                Messages:       Expected to find no apps but found 1 apps=[0x1400011e000]
--- FAIL: TestGetAppsEmpty (0.04s)
=== RUN   TestRouterNotRunning
    ddevapp_test.go:2889: 
                Error Trace:    /Users/stephen/Documents/git/dfm-ddev/pkg/ddevapp/ddevapp_test.go:2889
                Error:          Should not be: "ddev-router"
                Test:           TestRouterNotRunning
                Messages:       ddev-router was not supposed to be running but it was
--- FAIL: TestRouterNotRunning (0.01s)
```

```
Importing database via custom db_import_command
Failed to import database db for TestGitPull: Unable to validate import asset ~/tmp/ddev-pull-git-test-repo/db.sql.gz: invalid asset: file not found
    providerGit_test.go:67: 
                Error Trace:    /Users/stephen/tmp/ddevtest/TestGitPull905095413/providerGit_test.go:67
                Error:          Received unexpected error:
                                exit status 1
                Test:           TestGitPull
    providerGit_test.go:69: 
                Error Trace:    /Users/stephen/tmp/ddevtest/TestGitPull905095413/providerGit_test.go:69
                Error:          unable to find file "/Users/stephen/tmp/ddevtest/TestGitPull905095413/web/sites/default/files/tmp/veggie-pasta-bake-hero-umami.jpg"
                Test:           TestGitPull
ERROR 1146 (42S02) at line 1: Table 'db.users_field_data' doesn't exist
    providerGit_test.go:74: 
                Error Trace:    /Users/stephen/tmp/ddevtest/TestGitPull905095413/providerGit_test.go:74
                Error:          Received unexpected error:
                                ComposeCmd failed to run 'COMPOSE_PROJECT_NAME=ddev-TestGitPull docker-compose -f /Users/stephen/tmp/ddevtest/TestGitPull905095413/.ddev/.ddev-docker-compose-full.yaml exec -T db bash -c set -eu && ( echo 'select COUNT(*) from users_field_data where mail="margaret.hopper@example.com";' | mysql -N)', action='[exec -T db bash -c set -eu && ( echo 'select COUNT(*) from users_field_data where mail="margaret.hopper@example.com";' | mysql -N)]', err='exit status 1', stdout='', stderr='ERROR 1146 (42S02) at line 1: Table 'db.users_field_data' doesn't exist'
                Test:           TestGitPull
    providerGit_test.go:75: 
                Error Trace:    /Users/stephen/tmp/ddevtest/TestGitPull905095413/providerGit_test.go:75
                Error:          Should be true
                Test:           TestGitPull
Container ddev-TestGitPull-dba  Removed
Container ddev-TestGitPull-db  Removed
Container ddev-TestGitPull-web  Removed
Network ddev-testgitpull_default  Removed
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestGitPull-mariadb for project TestGitPull was deleted
Volume TestGitPull-postgres for project TestGitPull was deleted
Volume TestGitPull_project_mutagen for project TestGitPull was deleted
Deleted docker image drud/ddev-dbserver-mariadb-10.4:v1.21.4-TestGitPull-built
Deleted docker image drud/ddev-webserver:v1.21.4-TestGitPull-built
Project TestGitPull was deleted. Your code and configuration are unchanged.
--- FAIL: TestGitPull (18.11s)
```

**Comment by rfay** [comment]
The first one is counting running projects... and you have one running that skews it.

But TestGitPull is the one that failed in the tests right? On every platform? So step through it, do get set up so you can step through tests. It's worth the time.

When you want to run a single test (TestGitPull) the easiest way is to do it with your vscode, but you can also `make testpkg TESTARGS="-run TestGitPull"`

**Comment by rfay** [comment]
I'm happy to do a screenshare call with you and explore step-debugging and anything else you'd like to discuss. It's great to have you working on these things!

**Comment by DigitalFrontiersMedia** [comment]
> The first one is counting running projects... and you have one running that skews it.
> 

Ah. Didn't know that would be a factor.

> But TestGitPull is the one that failed in the tests right? On every platform? So step through it, do get set up so you can step through tests. It's worth the time.

I tried before to get Delve working, but there VSCode complains of errors in the launch.json file that is a part of the repo, and I find it hard to believe that something that enables step-debugging is committed as broken, and I obviously don't want to overwrite what must be working for others.  So I stopped there.

> 
> When you want to run a single test (TestGitPull) the easiest way is to do it with your vscode, but you can also `make testpkg TESTARGS="-run TestGitPull"`

Yes, I ran that before and get the same sort of response for the critical part.  Looks like everything goes south here:
`No DBPullCommand is defined for provider 'git'`
which comes form provider.go:336 in function getDatabaseBackups().

> I'm happy to do a screenshare call with you and explore step-debugging and anything else you'd like to discuss.

Yes, and maybe a general tour and orientation as well?  Available any time that works for you this week after today.

**Comment by rfay** [comment]
I can do Tuesday afternoon my time (US MT, UTC-7) or Wed morning. Let me know what works for you. Not sure I know your timezone? Are you already in Discord? https://discord.gg/hCZFfAMc5k

Also, if you haven't tried debugging in gitpod, please try it. For this PR all you have to do is go to https://gitpod.io/#https://github.com/drud/ddev/pull/4594 - that may give you a way  to compare your local vscode setup to what is working there.

You'll find I'm not very competent with vscode because I always use Goland and love it. But I'm sure we'll be able to get by.

What OS are you comfortable with, so I can use the same one?

**Comment by DigitalFrontiersMedia** [comment]
> I can do Tuesday afternoon my time (US MT, UTC-7) or Wed morning. Let me know what works for you. 

Tomorrow (Tuesday) works.  Whatever time works for you.

>Not sure I know your timezone? 

Eastern (in Sarasota, FL)

>Are you already in Discord? 

Yes, as DigitalFrontiersMedia

> Also, if you haven't tried debugging in gitpod, please try it. For this PR all you have to do is go to https://gitpod.io/#https://github.com/drud/ddev/pull/4594 - that may give you a way to compare your local vscode setup to what is working there.

Will have a look at that.

> What OS are you comfortable with, so I can use the same one?

I'm on MacOS Monterey (12.1)

**Comment by rfay** [comment]
Please PM me with your email in Discord and I'll send you a calendar invite.

**Comment by DigitalFrontiersMedia** [comment]
> > Also, if you haven't tried debugging in gitpod, please try it. For this PR all you have to do is go to https://gitpod.io/#https://github.com/drud/ddev/pull/4594 - that may give you a way to compare your local vscode setup to what is working there.
> 
> Will have a look at that.

Trying to run debugger in GitPod is giving exact same errors I get locally (screenshot attached).
Screen Shot 2023-01-30 at 2 51 22 PM

**Comment by rfay** [comment]
I rebased this so you'd have the new launch.json.

**Comment by rfay** [comment]
Rebased

**Comment by rfay** [comment]
Rebased. Any chance you can get this working before v1.21.5 @DigitalFrontiersMedia ?

**Comment by DigitalFrontiersMedia** [comment]
I've actually been coming to the conclusion that this is an issue/feature/bug of the yaml go pkg and how it reads/marshals the values.  There appears to be no way (yet) that I've seen to detect the difference between a string value that has been quoted and one that has not.  It seems like it might require reading the values differently because they both just get brought into memory as an unquoted string, i.e. the quotes are not included in the string, both with and without quotes is a valid string value upon read.  Still thinking about that.

**Comment by DigitalFrontiersMedia** [comment]
I may have an idea but still need to test if it's viable.

**Comment by DigitalFrontiersMedia** [comment]
Rebased.

**Comment by DigitalFrontiersMedia** [comment]
With the following 
```
environment_variables:
  test: variable with spaces
```
yaml should create a value of type string returned as:  variable with spaces

And that is valid.

With
```
environment_variables:
  test: "variable with spaces"
```
yaml is *forced* to create a value of type string returned as:  variable with spaces

And that is valid.

gopkg.in/yaml.v3 strips the quotes during unmarshalling except in situations where they are necessary to change the type of the value, e.g.:
```
environment_variables:
  test: 10
  testAsString: "10"
```
(First value interpreted as Integer, while second is forced to be a string and the quotes here change the type, whereas the type was not changed in the previous example).

The way the variables are injected/exported into the environment requires all of them to be output in a single line and forcing them to multiple export lines causes the injection command to break (probably because they are followed by piped commands).  So to prevent ambiguity between back-to-back keys and values, the values should enclosed in quotes to set them apart.

Unfortunately, environment_variables may be just about anything, including paths, which are stored as strings but should not be enclosed in quotes in order to prevent the home directory shortcut tilde from being taken as the literal name of a directory in a path.  There is no good way to differentiate between a string that contains slashes like "this/that or another" and a set of directories in a path like "this/that or another".

Note that spaces may also be in directory paths as above, which prevents that as being a determining factor in discriminating values for being enclosed in quotes or not.  Spaces in directory names would have an escape slash, but there's no guarantee you wouldn't have a string variable that has the same pattern, too.

Together, these issues preclude being able to discrminate which values to enclose in quotes after the ProviderInfo has already been Unmarshalled.

Based on this suggestion from InfoWorld that if you really want a string in yaml you should disambiguated it with quotes (see the section "When in doubt, quote strings" in "7 YAML Gotchas", [https://www.infoworld.com/article/3669238/7-yaml-gotchas-to-avoidand-how-to-avoid-them.html](https://www.infoworld.com/article/3669238/7-yaml-gotchas-to-avoidand-how-to-avoid-them.html)), and assuming no other ProviderInfo would follow the pattern:
`key: "value"`
I search for this pattern and enclose the quoted string in quotes itself so the first set are retained after unmarshalling as part of the string.

Another way I could think of would be to parse the values a second time using a package other than gopkg.in/yaml.v3 where quotes are preserved, compare the values to the unmarshalled values and swap as necessary.

The only other way would be to simply replace spaces in the values in injectedEnvironment() with escaped spaces, which should work for both paths and shell inline values.  Trying this next due to failures in tests with my attempt to alter loaded yaml prior to unmarshalling.

**Comment by DigitalFrontiersMedia** [comment]
I'm completely perplexed by this.  There are three failing checks on the latest attempt.  And oddly, if I just run `make test` on the unabridged `master` branch, I also get failures (albeit different ones).

Have any thoughts to enlighten me?

**Comment by rfay** [comment]
Thanks for all your work on this! 

I invited you to buildkite, so you'll be able to restart tests. 

* https://buildkite.com/drud/ddev-macos-mutagen/builds/3569#0186710d-c4e7-4995-bb73-c8e6720e34e8 was due to docker being broken on that machine. Docker Desktop is really flaky. tb-macos-amd64-1 - I restarted the test. 
* https://buildkite.com/drud/ddev-macos-use-nfsmount/builds/6825#0186710d-9701-4408-b82a-3bc3c5563215 was due to docker desktop hanging. I rebooted the machine and retried the test.
* https://buildkite.com/drud/ddev-windows-mutagen/builds/3556#0186710d-88a9-478c-bccc-0f17a088dfa1 was a timeout on windows-mutagen at 4 hours, which isn't uncommon. I restarted. Unfortunately I don't have access to that testbot to turn it off, etc.

I'm on vacation still through the end of the month, so don't always have availability to fix these immediately. But you now can at least restart or study the failures to see if they're relevant.

**Comment by rfay** [comment]
And of course I totally appreciate your focused analysis of all the issues above, just haven't had enough focused brainpower to study it yet.

**Comment by DigitalFrontiersMedia** [comment]
No worries.  Rest up.  Thanks, that info helps my sanity.  Will check out when I have some time next.

**Comment by rfay** [comment]
Rebased.

**Comment by rfay** [comment]
Rebased; make sure you update your remote.

**Comment by DigitalFrontiersMedia** [comment]
Have a little time later today and hope to use it to check into the buildkite tests that failed.

**Comment by rfay** [comment]
* Did you actually solve this in the end using just one additional space? Or have you actually not yet made progress? 
* One thing I should have mentioned long ago is that I'd like to completely lose the `environment_variables` in the provider and use the other techniques (config.yaml, global_config, .env, etc). I don't know why I ever did it this way. platform.yaml already works this way.

**Comment by rfay** [comment]
> Have a little time later today and hope to use it to check into the buildkite tests that failed.

With something like this, if buildkite tests failed but not the github actions test suite, it was just broken docker desktop stuff probably.

**Comment by rfay** [comment]
I do think that since this will probably be targeted for a major release (v1.22) we should consider removing the environment_variables section and just do a breaking change. We could notify users that the change was happening if we find environment_variables in the provider.

**Comment by DigitalFrontiersMedia** [comment]
> * Did you actually solve this in the end using just one additional space? Or have you actually not yet made progress?

I made what I thought was progress except for the BuildKite failures.  Just sitting down to look at the BuildKite (and one other failure as of recent rebase changes).

> * One thing I should have mentioned long ago is that I'd like to completely lose the `environment_variables` in the provider and use the other techniques (config.yaml, global_config, .env, etc). I don't know why I ever did it this way. platform.yaml already works this way.

Yes, I recall that.

> I do think that since this will probably be targeted for a major release (v1.22) we should consider removing the environment_variables section and just do a breaking change. We could notify users that the change was happening if we find environment_variables in the provider.

I think this current patch does the trick.  So maybe we add another issue to provide the environment_variables check to be worked on after this in prep for deprecation rather than a breaking change so soon?

**Comment by DigitalFrontiersMedia** [comment]
I can't seem to continue with BuildKite.  It says "Authorization Required" with a "Continue with Google" option to sign in, and when I put in my Google account info, it states: "The account you selected is not a GSuite hosted account. Please try again and select your DDEV GSuite account."

**Comment by rfay** [comment]
Looks to me like the test failures currently are just random failures and don't matter, but if you get me your email by PM or otherwise I'll add you to buildkite.

**Comment by DigitalFrontiersMedia** [comment]
The only tests now failing is a BuildKite test ("buildkite/wsl2-docker-desktop") and "Tests / tests (apache-fpm, apache-fpm) (pull_request)".  That second one appears to not be able to pass due to:

> Unable to get list of available services: GET https://api.github.com/search/repositories?q=topic:ddev-get+fork:true+org:ddev: 403 You have exceeded a secondary rate limit. Please wait a few minutes before you try again.

It is "TestCmdGet".  If I just run that test by itself on my local after merging updates, it passes:
```
(base) stephen@MacBook-Pro:dfm-ddev [14:14:21] (20230127_dfm_issue3499_parsing_provider_variables_with_spaces): make testpkg TESTARGS="-run TestCmdGet"
building .gotmp/bin/darwin_arm64/ddev from ./cmd/... ./pkg/...
export PATH="/Users/stephen/Documents/git/dfm-ddev/.gotmp/bin/darwin_arm64:$PATH" DDEV_NO_INSTRUMENTATION=true CGO_ENABLED=0 DDEV_BINARY_FULLPATH=/Users/stephen/Documents/git/dfm-ddev/.gotmp/bin/darwin_arm64/ddev; go test  -p 1 -timeout 4h -v -installsuffix static -ldflags " -extldflags -static -X 'github.com/ddev/ddev/pkg/versionconstants.DdevVersion=v1.21.6-20-g376f2cd8-dirty' -X 'github.com/ddev/ddev/pkg/versionconstants.SegmentKey=' " ./pkg/fileutil ./pkg/appports ./pkg/archive ./pkg/util ./pkg/globalconfig ./pkg/dockerutil ./pkg/ddevhosts ./pkg/output ./pkg/appimport ./pkg/styles ./pkg/version ./pkg/versionconstants ./pkg/exec ./pkg/nodeps ./pkg/testcommon ./pkg/netutil ./pkg/updatecheck -run TestCmdGet
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/fileutil       0.123s [no tests to run]
?       github.com/ddev/ddev/pkg/appports       [no test files]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/archive        (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/util   (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/globalconfig   (cached) [no tests to run]
v1.21.5: Pulling from drud/ddev-webserver
309bccac2f20: Pulling fs layer
309bccac2f20: Verifying Checksum
309bccac2f20: Download complete
309bccac2f20: Pull complete
Digest: sha256:111eb8761d2176c1e527e53dc0bc88a6ccf43079396f8613ceaeafbb7718d6af
Status: Downloaded newer image for drud/ddev-webserver:v1.21.5
docker.io/drud/ddev-webserver:v1.21.5
2023/03/12 14:15:00 StartContainer() at 2023-03-12 14:15:00.60717 -0400 EDT m=+31.367747585
2023/03/12 14:15:00 ContainerWait at 2023-03-12 14:15:00.842457 -0400 EDT m=+31.603034918
2023/03/12 14:15:06 ContainerWait returrned at 2023-03-12 14:15:06.360626 -0400 EDT m=+37.121209585 out='phpstatus: OK /var/www/html: OK mailhog: OK ' err='<nil>'
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/dockerutil     37.427s [no tests to run]
?       github.com/ddev/ddev/pkg/ddevhosts      [no test files]
?       github.com/ddev/ddev/pkg/output [no test files]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/appimport      0.179s [no tests to run]
?       github.com/ddev/ddev/pkg/styles [no test files]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/version        0.119s [no tests to run]
?       github.com/ddev/ddev/pkg/versionconstants       [no test files]
?       github.com/ddev/ddev/pkg/exec   [no test files]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/nodeps 0.080s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/testcommon     0.123s [no tests to run]
?       github.com/ddev/ddev/pkg/netutil        [no test files]
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/updatecheck    0.142s [no tests to run]
export PATH="/Users/stephen/Documents/git/dfm-ddev/.gotmp/bin/darwin_arm64:$PATH" DDEV_NO_INSTRUMENTATION=true CGO_ENABLED=0 DDEV_BINARY_FULLPATH=/Users/stephen/Documents/git/dfm-ddev/.gotmp/bin/darwin_arm64/ddev; go test  -p 1 -timeout 4h -v -installsuffix static -ldflags " -extldflags -static -X 'github.com/ddev/ddev/pkg/versionconstants.DdevVersion=v1.21.6-20-g376f2cd8-dirty' -X 'github.com/ddev/ddev/pkg/versionconstants.SegmentKey=' " ./pkg/ddevapp -run TestCmdGet
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
could not remove volume TestPkgWordpress-mariadb: Docker volume 'TestPkgWordpress-mariadb' is in use by one or more containers and cannot be removed. Use 'docker rm -f priceless_bell' to remove them
could not remove volume TestPkgWordpress-postgres: Docker volume 'TestPkgWordpress-postgres' is in use by one or more containers and cannot be removed. Use 'docker rm -f priceless_bell' to remove them
Volume TestPkgWordpress_project_mutagen for project TestPkgWordpress was deleted
Project TestPkgWordpress was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgWordpress/TestPkgWordpress_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgWordpress3244332412

Copying complete
time="2023-03-12T14:15:27-04:00" level=error msg="TestMain startup: Failed to delete volume TestPkgWordpress-mariadb: Docker volume 'TestPkgWordpress-mariadb' is in use by one or more containers and cannot be removed. Use 'docker rm -f priceless_bell' to remove them"
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgDrupal8-mariadb for project TestPkgDrupal8 was deleted
Volume TestPkgDrupal8-postgres for project TestPkgDrupal8 was deleted
Volume TestPkgDrupal8_project_mutagen for project TestPkgDrupal8 was deleted
Project TestPkgDrupal8 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgDrupal8/TestPkgDrupal8_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgDrupal82585265291

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgDrupal7-mariadb for project TestPkgDrupal7 was deleted
Volume TestPkgDrupal7-postgres for project TestPkgDrupal7 was deleted
Volume TestPkgDrupal7_project_mutagen for project TestPkgDrupal7 was deleted
Project TestPkgDrupal7 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgDrupal7/TestPkgDrupal7_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgDrupal72441524026

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgDrupal6-mariadb for project TestPkgDrupal6 was deleted
Volume TestPkgDrupal6-postgres for project TestPkgDrupal6 was deleted
Volume TestPkgDrupal6_project_mutagen for project TestPkgDrupal6 was deleted
Project TestPkgDrupal6 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgDrupal6/TestPkgDrupal6_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgDrupal6199377597

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgBackdrop-mariadb for project TestPkgBackdrop was deleted
Volume TestPkgBackdrop-postgres for project TestPkgBackdrop was deleted
Volume TestPkgBackdrop_project_mutagen for project TestPkgBackdrop was deleted
Project TestPkgBackdrop was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgBackdrop/TestPkgBackdrop_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgBackdrop1680304604

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgTypo3-mariadb for project TestPkgTypo3 was deleted
Volume TestPkgTypo3-postgres for project TestPkgTypo3 was deleted
Volume TestPkgTypo3_project_mutagen for project TestPkgTypo3 was deleted
Project TestPkgTypo3 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgTypo3/TestPkgTypo3_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgTypo33467802536

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume testpkgmagento-mariadb for project testpkgmagento was deleted
Volume testpkgmagento-postgres for project testpkgmagento was deleted
Volume testpkgmagento_project_mutagen for project testpkgmagento was deleted
Project testpkgmagento was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/testpkgmagento/testpkgmagento_siteArchive to /Users/stephen/tmp/ddevtest/testpkgmagento1347810154

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume testpkgmagento2-mariadb for project testpkgmagento2 was deleted
Volume testpkgmagento2-postgres for project testpkgmagento2 was deleted
Volume testpkgmagento2_project_mutagen for project testpkgmagento2 was deleted
Project testpkgmagento2 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/testpkgmagento2/testpkgmagento2_siteArchive to /Users/stephen/tmp/ddevtest/testpkgmagento21585978640

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgDrupal9-mariadb for project TestPkgDrupal9 was deleted
Volume TestPkgDrupal9-postgres for project TestPkgDrupal9 was deleted
Volume TestPkgDrupal9_project_mutagen for project TestPkgDrupal9 was deleted
Project TestPkgDrupal9 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgDrupal9/TestPkgDrupal9_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgDrupal93318391035

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgLaravel-mariadb for project TestPkgLaravel was deleted
Volume TestPkgLaravel-postgres for project TestPkgLaravel was deleted
Volume TestPkgLaravel_project_mutagen for project TestPkgLaravel was deleted
Project TestPkgLaravel was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgLaravel/TestPkgLaravel_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgLaravel3814701223

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume testpkgshopware6-mariadb for project testpkgshopware6 was deleted
Volume testpkgshopware6-postgres for project testpkgshopware6 was deleted
Volume testpkgshopware6_project_mutagen for project testpkgshopware6 was deleted
Project testpkgshopware6 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/testpkgshopware6/testpkgshopware6_siteArchive to /Users/stephen/tmp/ddevtest/testpkgshopware63919756079

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgPHP-mariadb for project TestPkgPHP was deleted
Volume TestPkgPHP-postgres for project TestPkgPHP was deleted
Volume TestPkgPHP_project_mutagen for project TestPkgPHP was deleted
Project TestPkgPHP was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgPHP/TestPkgPHP_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgPHP1166036784

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgDrupal10-mariadb for project TestPkgDrupal10 was deleted
Volume TestPkgDrupal10-postgres for project TestPkgDrupal10 was deleted
Volume TestPkgDrupal10_project_mutagen for project TestPkgDrupal10 was deleted
Project TestPkgDrupal10 was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgDrupal10/TestPkgDrupal10_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgDrupal102205841860

Copying complete
Not trying to remove hostnames because DDEV_NONINTERACTIVE=true
Volume TestPkgCraftCms-mariadb for project TestPkgCraftCms was deleted
Volume TestPkgCraftCms-postgres for project TestPkgCraftCms was deleted
Volume TestPkgCraftCms_project_mutagen for project TestPkgCraftCms was deleted
Project TestPkgCraftCms was deleted. Your code and configuration are unchanged.
Copying directory /Users/stephen/.ddev/testcache/TestPkgCraftCms/TestPkgCraftCms_siteArchive to /Users/stephen/tmp/ddevtest/TestPkgCraftCms3320256671

Copying complete
testing: warning: no tests to run
PASS
ok      github.com/ddev/ddev/pkg/ddevapp        107.253s [no tests to run]
```

**Comment by DigitalFrontiersMedia** [comment]
Holy moly, all checks have passed!

I'd be grateful, Admiral, if you'd give the word.

**Comment by rfay** [comment]
Congratulations!

**Review by rfay — APPROVED** [review]
This looks like the perfect simple solution to me. Tested with your manual testing instructions and it looks fine. The one comment was to use the simpler approach to adding a literal backslash using backticks. You can just commit that.

**Comment by rfay** [comment]
Just waiting for that one commit above if you approve it @DigitalFrontiersMedia .

File: pkg/ddevapp/provider.go
Comment (DigitalFrontiersMedia) [reply pkg/ddevapp/provider.go:442]: Ah!  Thank you!  Perfect.

**Comment by DigitalFrontiersMedia** [comment]
I hope I did that right.  I didn't see an explicit "approval" button.

**Pull request merged**
